rust/connlib/tunnel/src/lib.rs

```diff
@@ -554,3 +505,244 @@ where
+                let dev = Arc::clone(self);
+
+                let mut connected_gateway_ids: Vec<_> = dev
+                    .gateway_awaiting_connection
+                    .lock()
+                    .clone()
+                    .into_keys()
+                    .collect();
+                connected_gateway_ids
+                    .extend(dev.resources_gateways.lock().values().collect::<Vec<_>>());
+                tracing::trace!(
+                    gateways = ?connected_gateway_ids,
+                    "connected_gateways"
+                );
+                tokio::spawn(async move {
+                    let mut interval = tokio::time::interval(MAX_SIGNAL_CONNECTION_DELAY);
+                    loop {
+                        interval.tick().await;
+                        let reference = {
+                            let mut awaiting_connections = dev.awaiting_connection.lock();
+                            let Some(awaiting_connection) =
+                                awaiting_connections.get_mut(&ConnId::from(resource.id()))
+                            else {
+                                break;
+                            };
+                            if awaiting_connection.response_received {
+                                break;
+                            }
+                            awaiting_connection.total_attemps += 1;
+                            awaiting_connection.total_attemps
+                        };
+                        if let Err(e) = dev
+                            .control_signaler
+                            .signal_connection_to(&resource, &connected_gateway_ids, reference)
+                            .await
+                        {
+                            // Not a deadlock because this is a different task
+                            dev.awaiting_connection.lock().remove(&conn_id);
+                            tracing::error!(error = ?e, "start_resource_connection");
+                            let _ = dev.callbacks.on_error(&e);
+                        }
+                    }
+                });
+            }
+        }
+    }
+}
+
+impl<C, CB> Tunnel<C, CB, GatewayState>
+where
+    C: ControlSignal + Send + Sync + 'static,
+    CB: Callbacks + 'static,
+{
+    /// Sets the interface configuration and starts background tasks.
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub async fn set_interface(self: &Arc<Self>, config: &InterfaceConfig) -> Result<()> {
+        let device = create_iface(config, self.callbacks()).await?;
+        *self.device.write().await = Some(device.clone());
+
+        self.start_timers().await?;
+        self.start_device(device);
+
+        tracing::debug!("background_loop_started");
+
+        Ok(())
+    }
+
+    fn start_device(self: &Arc<Self>, mut device: Device) {
+        let tunnel = Arc::clone(self);
+
+        *self.iface_handler_abort.lock() = Some(
+            tokio::spawn(async move {
```

Comment:
For both `start_device` I'd like if we can move closure to its own function? I find it easier to read.